File: src/ggrc/models/event.py
# Copyright (C) 2019 Google Inc.
# Licensed under http://www.apache.org/licenses/LICENSE-2.0 <see LICENSE file>

from sqlalchemy import orm

from ggrc import db
from ggrc.models.mixins import Base
from ggrc.models import reflection


class Event(Base, db.Model):
  __tablename__ = 'events'

  action = db.Column(
      db.Enum(u'POST', u'PUT', u'DELETE', u'BULK', u'GET'),
      nullable=False,
  )
  resource_id = db.Column(db.Integer)
  resource_type = db.Column(db.String)

  revisions = db.relationship(
      'Revision',
      backref='event',
      cascade='all, delete-orphan',
  )

  _api_attrs = reflection.ApiAttributes(
      'action',
      'resource_id',
      'resource_type',
      'revisions',
  )

  _include_links = [
      'revisions',
  ]

  @staticmethod
  def _extra_table_args(class_):
    return (
        db.Index('events_modified_by', 'modified_by_id'),
    )

  @classmethod
  def eager_query(cls, **kwargs):
    query = super(Event, cls).eager_query(**kwargs)
    return query.options(
        orm.subqueryload('revisions').undefer_group('Revision_complete'),
    )

File: src/ggrc-client/js/models/service-models/event.js
/*
    Copyright (C) 2019 Google Inc.
    Licensed under http://www.apache.org/licenses/LICENSE-2.0 <see LICENSE file>
*/

import Cacheable from '../cacheable';
import Stub from '../stub';

export default Cacheable.extend({
  root_object: 'event',
  root_collection: 'events',
  findAll: 'GET /api/events',
  list_view_options: {
    find_params: {
      __include: 'revisions',
    },
  },
  attributes: {
    modified_by: Stub,
  },
}, {});
